elastic/beats
# x-pack/filebeat/input/http_endpoint: Add CRC validation for webhooks

## What does this PR do?

Added support to process CRC requests in the filebeat's http_endpoint input.

Currently, the only module that we support with CRC webhook validation is Zoom, so it has been added the procedure to process Zoom CRC requests according to [its documentation](https://developers.zoom.us/docs/api/rest/webhook-reference/#validate-your-webhook-endpoint).

It has been added a new setting `crc.provider` to the `http_endpoint` configuration to choose the webhook provider in case the list of supported services is in the future.

## Why is it important?


Currently, the HTTP endpoint can only send static responses. Some use cases such as the CRC validation require dynamic responses based on the content of the request or other data.

## Checklist




## How to test this PR locally

For now, the only allowed provider for CRC requests is [Zoom](https://developers.zoom.us/docs/api/rest/webhook-reference/#validate-your-webhook-endpoint), so we have to add the `crc.provider` setting as follows in the `filebeat.yml` file:
```yml
- type: http_endpoint
  enabled: true
  listen_address: localhost
  listen_port: 5066
  url: "/test/"
  response_code: 200
  response_body: '{"message": "test success"}'
  secret.header: "secretHeaderTest"
  secret.value: "secretValueTest"
  crc.provider: "Zoom"
```

It is also needed to have the webhook configured with a secret token, for this purpose, options `secret.header` and `secret.value` should be set.

Once filebeat is running, we can send HTTP requests to verify that it behaves as expected.

**Regular event**

```
# curl -X POST http://localhost:5066/test/ -H 'Content-Type: application/json' -H 'secretHeaderTest: secretValueTest' -d '{"message":"test_event"}' -v
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying 127.0.0.1:5066...
* Connected to localhost (127.0.0.1) port 5066 (#0)
> POST /test/ HTTP/1.1
> Host: localhost:5066
> User-Agent: curl/7.87.0
> Accept: */*
> Content-Type: application/json
> secretHeaderTest: secretValueTest
> Content-Length: 24
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Content-Type: application/json
< Date: Thu, 27 Apr 2023 12:34:03 GMT
< Content-Length: 27
<
* Connection #0 to host localhost left intact
{"message": "test success"}#
```

For this request, we can see the response is the expected (the configured one), we also check that the event is generated and sent:

```js
{
  "@timestamp": "2023-04-27T12:34:03.228Z",
  "@metadata": {
    "beat": "filebeat",
    "type": "_doc",
    "version": "8.8.0"
  },
  "json": {
    "message": "test_event"
  },
  "input": {
    "type": "http_endpoint"
  },
  "ecs": {
    "version": "8.0.0"
  },
...
```

**CRC request**

```
# curl -X POST http://localhost:5066/test/ -H 'Content-Type: application/json' -H 'secretHeaderTest: secretValueTest' -d '{"event_ts":1654503849680,"event":"endpoint.url_validation","payload":{"plainToken":"qgg8vlvZRS6UYooatFL8Aw"}}' -v
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying 127.0.0.1:5066...
* Connected to localhost (127.0.0.1) port 5066 (#0)
> POST /test/ HTTP/1.1
> Host: localhost:5066
> User-Agent: curl/7.87.0
> Accept: */*
> Content-Type: application/json
> secretHeaderTest: secretValueTest
> Content-Length: 110
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Content-Type: application/json
< Date: Thu, 27 Apr 2023 12:36:00 GMT
< Content-Length: 123
<
* Connection #0 to host localhost left intact
{"encryptedToken":"70c1f2e2e6ca2d39297490d1f9142c7d701415ea8e6151f6562a08fa657a40ff","plainToken":"qgg8vlvZRS6UYooatFL8Aw"}#
```

In this case, when sending a CRC request event to the server, the response message contains the hash of the challenge token. In addition, no event is generated.

## Related issues

Closes https://github.com/elastic/beats/issues/35100
